tests/connect_redirect/connect_redirect_tests.cpp

```diff
@@ -0,0 +1,433 @@
+    IN6ADDR_SETLOOPBACK((PSOCKADDR_IN6)&_globals.addresses[socket_family_t::IPv6].loopback_address);
+
+    _globals.destination_port = _destination_port;
+    _globals.proxy_port = _proxy_port;
+    _globals_initialized = true;
+}
+
+static void
+_load_and_attach_ebpf_programs(_Outptr_ struct bpf_object** return_object)
+{
+    struct bpf_object* object = bpf_object__open("cgroup_sock_addr2.o");
+    REQUIRE(object != nullptr);
+    REQUIRE(bpf_object__load(object) == 0);
+
+    bpf_program* connect_program_v4 = bpf_object__find_program_by_name(object, "connect_redirect4");
+    REQUIRE(connect_program_v4 != nullptr);
+
+    int result = bpf_prog_attach(
+        bpf_program__fd(const_cast<const bpf_program*>(connect_program_v4)), 0, BPF_CGROUP_INET4_CONNECT, 0);
+    REQUIRE(result == 0);
+
+    bpf_program* connect_program_v6 = bpf_object__find_program_by_name(object, "connect_redirect6");
+    REQUIRE(connect_program_v6 != nullptr);
+
+    result = bpf_prog_attach(
+        bpf_program__fd(const_cast<const bpf_program*>(connect_program_v6)), 0, BPF_CGROUP_INET6_CONNECT, 0);
+    REQUIRE(result == 0);
+
+    *return_object = object;
+}
+
+static void
+_update_policy_map(
+    _In_ const struct bpf_object* object,
+    _In_ sockaddr_storage& destination,
+    _In_ sockaddr_storage& proxy,
+    uint16_t destination_port,
+    uint16_t proxy_port,
+    uint32_t protocol,
+    bool dual_stack,
+    bool add)
+{
+    bpf_map* policy_map = bpf_object__find_map_by_name(object, "policy_map");
+    REQUIRE(policy_map != nullptr);
+
+    fd_t map_fd = bpf_map__fd(policy_map);
+
+    // Insert / delete redirect policy entry in the map.
+    destination_entry_t key = {0};
+    destination_entry_t value = {0};
+
+    if (_globals.family == AF_INET) {
+        const uint8_t* v4_destination = nullptr;
+        const uint8_t* v4_proxy = nullptr;
+
+        if (dual_stack) {
+            struct sockaddr_in6* v6_destination = (struct sockaddr_in6*)&destination;
+            struct sockaddr_in6* v6_proxy = (struct sockaddr_in6*)&proxy;
+
+            v4_destination = IN6_GET_ADDR_V4MAPPED((IN6_ADDR*)&v6_destination->sin6_addr);
+            v4_proxy = IN6_GET_ADDR_V4MAPPED((IN6_ADDR*)&v6_proxy->sin6_addr);
+        } else {
+            v4_destination = (UCHAR*)&((struct sockaddr_in*)&destination)->sin_addr.S_un.S_addr;
+            v4_proxy = (UCHAR*)&((struct sockaddr_in*)&proxy)->sin_addr.S_un.S_addr;
+        }
+
+        key.destination_ip.ipv4 = *((uint32_t*)v4_destination);
+        value.destination_ip.ipv4 = *((uint32_t*)v4_proxy);
+    } else {
+        struct sockaddr_in6* v6_destination = (struct sockaddr_in6*)&destination;
+        struct sockaddr_in6* v6_proxy = (struct sockaddr_in6*)&proxy;
+        memcpy(key.destination_ip.ipv6, v6_destination->sin6_addr.u.Byte, sizeof(key.destination_ip.ipv6));
+        memcpy(value.destination_ip.ipv6, v6_proxy->sin6_addr.u.Byte, sizeof(value.destination_ip.ipv6));
+    }
```

Comment:
Seems like this block can be collapsed.  if dual_stack then use what you have, else the other two cases can be combined into just
```
memcpy(&key.destination, INETADDR_ADDRESS(&destination), INET_ADDR_LENGTH(_globals.family));
memcpy(&value.destination, INETADDR_ADDRESS(&proxy), INET_ADDR_LENGTH(_globals.family));

INET_SET_ADDRESS(_globals.family, &key.destination_ip, 
```

Reply:
Fixed.